source/common/common/containers.h

```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+
+namespace Envoy {
+namespace Common {
+
+/**
+ * Invokes a function for all elements in a container that accepts a completion callback which will
+ * be invoked asynchronously, invoking a final completion callback once all callbacks associated
+ * with elements have been invoked.
+ */
+template <class ContainerT, class UpdateCbT>
+void applyToAllWithCompletionCallback(const ContainerT& container, UpdateCbT update_cb,
+                                      std::function<void()> done_cb) {
+
+  auto remaining_elements = std::make_shared<uint64_t>(container.size());
+  for (auto element : container) {
+    update_cb(element, [remaining_elements, done_cb] {
```

Comment:
So, before we were doing N threads * M providers, but now we just do M providers. Presumably there was something hiding underneath the provider `onConfigUpdate` implying N concurrency. How does this now go away?

In general, this continuation passing style is really hard to read in C++ (I know it's common in some languages). @kyessenov is there a way to simplify the interfaces here to make them less hard to reason about in terms of concurrency?

Reply:
Previously it would be decremented on the per thread callback, but I changed it to be done on the main thread completion callback instead. This eliminated the need to check the concurrency and also made the code no longer run on different threads, removing the need for atomics.

Reply:
It might be possible to do something clever around RAII semantics and a shared pointer to avoid some of this complexity, let me try some stuff out.